klippy/extras/bed_mesh.py

```diff
@@ -150,38 +153,67 @@ def __init__(self, config, bedmesh):
+
         x_cnt, y_cnt = parse_pair(
             config, ('probe_count', '3'), check=False, cast=int, minval=3)
         self.probe_params['x_count'] = x_cnt
         self.probe_params['y_count'] = y_cnt
-        min_x, min_y = parse_pair(config, ('min_point',))
-        max_x, max_y = parse_pair(config, ('max_point',))
-        if max_x <= min_x or max_y <= min_y:
-            raise config.error('bed_mesh: invalid min/max points')
-        x_dist = (max_x - min_x) / (x_cnt - 1)
-        y_dist = (max_y - min_y) / (y_cnt - 1)
-        # floor distances down to next hundredth
-        x_dist = math.floor(x_dist * 100) / 100
-        y_dist = math.floor(y_dist * 100) / 100
-        if x_dist <= 1. or y_dist <= 1.:
-            raise config.error("bed_mesh: min/max points too close together")
-        # re-calc x_max
-        max_x = min_x + x_dist * (x_cnt - 1)
-        pos_y = min_y
-        points = []
-        for i in range(y_cnt):
-            for j in range(x_cnt):
-                if not i % 2:
-                    # move in positive directon
-                    pos_x = min_x + j * x_dist
-                else:
-                    # move in negative direction
-                    pos_x = max_x - j * x_dist
-                points.append((pos_x, pos_y))
-            pos_y += y_dist
+
+        if self._bed_shape == 'square':
+            min_x, min_y = parse_pair(config, ('min_point',))
+            max_x, max_y = parse_pair(config, ('max_point',))
+            if max_x <= min_x or max_y <= min_y:
+                raise config.error('bed_mesh: invalid min/max points')
+            x_dist = (max_x - min_x) / (x_cnt - 1)
+            y_dist = (max_y - min_y) / (y_cnt - 1)
+            # floor distances down to next hundredth
+            x_dist = math.floor(x_dist * 100) / 100
+            y_dist = math.floor(y_dist * 100) / 100
+            if x_dist <= 1. or y_dist <= 1.:
+                raise config.error("bed_mesh: min/max points too close together")
+            # re-calc x_max
+            max_x = min_x + x_dist * (x_cnt - 1)
+            pos_y = min_y
+
+            for i in range(y_cnt):
+                for j in range(x_cnt):
+                    if not i % 2:
+                        # move in positive directon
+                        pos_x = min_x + j * x_dist
+                    else:
+                        # move in negative direction
+                        pos_x = max_x - j * x_dist
+                    points.append((pos_x, pos_y))
+                pos_y += y_dist
+
+        elif self._bed_shape == 'circle':
+            self._radius = config.getfloat('radius', 50.)
+
+            self._distance_y = ((self._radius * 2) / (y_cnt - 1))
+            self._distance_x = ((self._radius * 2) / (x_cnt - 1))
+
+            for i in range(y_cnt):
+                pos_y = -self._radius + self._distance_y * i
+                for j in range(x_cnt):
+                    pos_x = -self._radius + self._distance_x * j
```

Comment:
In my original comment, where I referred to "excess  toolhead movement", this is the location I was speaking of. While it is true that the tool will move in a linear fashion in the x direction, when the tool reaches the last x coordinate at a given y coordinate, it will then move to the other side of the bed and begin probing again.  It would be relatively easy to convert this to a zigzag pattern by adding the following:
```
if i & 1 == 0:
    # move in positive direction
    pos_x = -radius + j * x_dist
else:
    # move in negative direction
    pos_x = radius - j * x_dist
```
Another thing to keep in mind is that for interpolation purposes, we probably want a minimum of 3 probed points along each axis (thus x_count and y_count options should be minimums for round beds, not maximums).  This makes the existing interpolation algorithms work correctly, although it would be possible to simply use linear interpolation for 2 points.  What we don't want are single coordinates by themselves at the outer radius.